plugin/handlergen.go

```diff
@@ -118,23 +118,71 @@ func (p *OrmPlugin) generateReadHandler(message *generator.Descriptor) {
 
 func (p *OrmPlugin) generateApplyFieldMask(message *generator.Descriptor) {
 	typeName := p.TypeName(message)
-
 	p.P(`// DefaultApplyFieldMask`, typeName, ` patches an pbObject with patcher according to a field mask.`)
 	p.P(`func DefaultApplyFieldMask`, typeName, `(ctx context.Context, patchee *`,
 		typeName, `,ormObj *`, typeName, `ORM, patcher *`,
 		typeName, `, updateMask *`, p.Import(fmImport),
-		`.FieldMask, db *`, p.Import(gormImport), `.DB) (*`, typeName, `, error) {`)
-	p.P(`var err error`)
+		`.FieldMask, prefix string, db *`, p.Import(gormImport), `.DB) (*`, typeName, `, error) {`)
 
+	p.P(`if patcher == nil {`)
+	p.P(`return nil, nil`)
+	p.P(`} else if patchee == nil || ormObj == nil {`)
+	p.P(`return nil, errors.New("Patchee and ormObj inputs to DefaultApplyFieldMask`,
+		typeName, ` must be non-nil")`)
+	p.P(`}`)
+	p.P(`var err error`)
+	hasNested := false
+	for _, field := range message.GetField() {
+		fieldType, _ := p.GoType(message, field)
+		if field.IsMessage() && p.isOrmable(generator.CamelCase(fieldType)) && !field.IsRepeated() {
+			p.P(`var updated`, generator.CamelCase(field.GetName()), ` bool`)
+			hasNested = true
+		}
+	}
 	// Patch pbObj with input according to a field mask.
-	p.P(`for _, f := range updateMask.GetPaths() {`)
+	if hasNested {
+		p.P(`for i, f := range updateMask.Paths {`)
+	} else {
+		p.P(`for _, f := range updateMask.Paths {`)
+	}
 	for _, field := range message.GetField() {
 		ccName := generator.CamelCase(field.GetName())
-		p.P(`if f == "`, ccName, `" {`)
-		p.P(`patchee.`, ccName, ` = patcher.`, ccName)
-		p.removeChildAssociationsByName(message, ccName)
-		p.setupOrderedHasManyByName(message, ccName)
-		p.P(`}`)
+		fieldType, _ := p.GoType(message, field)
+		//  for ormable message, do recursive patching
+		if field.IsMessage() && p.isOrmable(fieldType) && !field.IsRepeated() {
+			p.UsingGoImports("strings")
+			p.P(`if strings.HasPrefix(f, prefix+"`, ccName, `.") && !updated`, ccName, ` {`)
```

Comment:
say we have the following:
```
struct {
  Foo struct {
    Bar struct {
      Baz string
    }
  }
}
```
and FieldMask with path `Foo`

Will this catch it? I see the dot at the end which would imply that a full replacement of `Foo` would get missed

Reply:
Ah. You are correct.